File: lemon.pl
use strict;
use warnings;
use Memoize;

memoize('calc');

#use bignum;

my $input = <>;
chomp $input;
$input =~ /\s*(\d+)\s+(\d+)/;

my ($total, $max) = ($1, $2);


for (my $i = 0; $i <= $total; $i++) {
    for (my $j = 0; $j <= $max; $j++) {
        calc($j, $i);
    }
}

my @cache;

for (my $i  = 0; $i <= $max; $i++) {
    $cache[0][$i] = 1;
}

#$cache[0][$max+1]=0;

for (my $len = 1; $len <= $total; $len++) {
    for (my $lemons = 0; $lemons <= $max; $lemons++) {
        
    $cache[$len][$lemons] = 0;
        if ($lemons < $max) {
            $cache[$len][$lemons] = $cache[$len - 1][$lemons +1];
        }

        $cache[$len][$lemons] +=  $cache[$len - 1][0];
    }
}

use Data::Dumper;

#warn Dumper(\@cache);

warn $cache[$total][0];

my $banan = kbanacci($total+1, $max+1);

warn Dumper($banan);

#my $result = calc(0, $total);

#warn $result;

sub calc
{
    my ($prev_lemons, $total_len, $d) = @_;

    $d //=0;

    if ($total_len == 0) {
        return 1;
    }
    
    my $rc = 0;

    if ($prev_lemons < $max) {
#        warn "".(" "x$d)."using: ".($prev_lemons + 1)." ".($total_len - 1);
        $rc += calc($prev_lemons + 1, $total_len - 1, $d+1);
    }

#    warn "".(" "x$d)."using: ".(0)." ".($total_len - 1);
    $rc += calc(0, $total_len - 1, $d+1);
    
    return $rc;
}

#Возвращает n-тое число k-баначи

sub kbanacci
{
    my ($n, $k) = @_;

    my @cache;

    $cache[0] = 1;
    $cache[1] = 1;
    

    if ($k > 1) {
        $cache[2] = 2;

        for (my $i = 3; $i <= $k; $i++) {
            $cache[$i] = $cache[$i-1]*2;
        }
    }

    for (my $i = 0; $i < $n; $i++) {
#        warn Dumper(\@cache);
        my $x = $cache[$k] - $cache[0] + $cache[$k];
        
        for (my $j = 1; $j <= $k; $j++) {
            $cache[$j - 1] = $cache[$j];
        }

        $cache[$k] = $x;
    }
    
    return \@cache;
}

File: lemon.c
#include "bignum.h"

int main(int argc, char** argv) 
{
    int total;
    int max;

    scanf("%d%d", &total, &max);
    
    long_number_t *cache[2][max+1];

    for (int i = 0; i < 2; i++) {
        for (int j = 0; j < max+1; j++) {
            cache[i][j] = new_number();
        }
    }

    for (int i  = 0; i <= max; i++) {
        *(cache[0][i]->values) = 1;
    }
    
    int prev = 0;

    for ( int len = 1; len <= total; len++) {
        int cur = (prev + 1) % 2;

        for (int lemons = 0; lemons <= max; lemons++) {
            zero_number(cache[cur][lemons]);
        
            if (lemons < max) {
                add_long_number(cache[cur][lemons], cache[prev][lemons +1]);
            }

            add_long_number(cache[cur][lemons],  cache[prev][0]);
        }

        prev = (prev +1) % 2;
    }

    print_long_number(cache[prev][0]);
}

long_number_t *kbanacci(int n, int k) 
{
    
    long_number_t *cache = (long_number_t *) malloc(sizeof(long_number_t*)*($k+1));
    
    long_number_t *two = new_number;
    two->values[0] = 2;
    
    cache[0] = new_number();
    *(cache[0]->values)=2;

    for (int i = 1; i <= k ; i++) {
        *(cache + i) =  multiply_long_numbers((cache+i-1), two);
    }

    for (int i = 0; i < n; i++) {
        my $x = $cache[$k] - $cache[0] + $cache[$k];
        
        for (my $j = 1; $j <= $k; $j++) {
            $cache[$j - 1] = $cache[$j];
        }

        $cache[$k] = $x;
    }
    
    return \@cache;
}
